File: rust/leet_code/src/2958.rs
use std::collections::HashMap;

impl Solution {
    pub fn max_subarray_length(nums: Vec<i32>, k: i32) -> i32 {
        let mut left = 0;
        let mut counter: HashMap<i32, i32> = HashMap::new();
        let mut max_length = 0;
        for right in 0..nums.len() {
            let num = nums[right];
            let count = counter.entry(num).or_insert(0);
            *count += 1;
            while *counter.get(&num).unwrap() > k {
                *counter.get_mut(&nums[left]).unwrap() -= 1;
                if *counter.get(&nums[left]).unwrap() == 0 {
                    counter.remove(&nums[left]);
                }
                left += 1;
            }
            max_length = max_length.max(right - left + 1);
        }
        max_length as i32
    }
}

File: swift/leet_code/2958.swift
class Solution {
    func maxSubarrayLength(_ nums: [Int], _ k: Int) -> Int {
        var left = 0
        var counter = [Int: Int]()
        var max_length = 0
        for right in 0..<nums.count {
            counter[nums[right], default: 0] += 1
            while counter[nums[right], default: 0] > k {
                counter[nums[left], default: 0] -= 1
                if counter[nums[left], default: 0] == 0 {
                    counter.removeValue(forKey: nums[left])
                }
                left += 1
            }
            max_length = max(max_length, right - left + 1)
        }
        return max_length
    }
}